### PR TITLE
bpf: Add EarlyTraceNotification option

### PR DESCRIPTION
**Pull request by joestringer**
Add an option to enable (default off) early trace events when a packet
is received by Cilium.

Related: #4080

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4763)

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by tgraf** [comment]
Change looks good but I'm not sure we can merge this without having the aggregation in place as we still need to provide the same L3/L4 flows behavior but aggregated.

File: test/runtime/monitor.go [hunk 1]
```diff
@@ -36,6 +36,10 @@ const (
 	// value for the Cilium monitor
 	MonitorTraceNotification = "TraceNotification"
 
+	// MonitorEarlyTrace represents the EarlyTraceNotification configuration
```
Comment (ianvernon) [inline test/runtime/monitor.go:39]: Nit: this definition is somewhat circular; i.e., what differentiates an early trace notification vs. a trace notification? It's not obvious without looking at the code itself.

**Review by ianvernon — APPROVED** [review]
LGTM

File: test/runtime/monitor.go (hunk 1 above)
Comment (joestringer) [reply test/runtime/monitor.go:39]: Yeah, this comment doesn't provide much helpful context but it's in line with each of the comments above the other configuration options declared in this file. Maybe we can do better at the actual option definition location.

**Comment by joestringer** [comment]
@tgraf The intended aggregation will cause only a maximum of a single event to be fired (maybe no event) based on a packet, however then we have to decide whether to fire an event when the packet is received or when the packet has been accepted (ie, transmitted to the next step, that is, `TraceTo*`). I'm assuming that with aggregation, there will no longer be any need for triggering events based on packet receive.

**Comment by joestringer** [comment]
Upon further thought, I'm considering changing this up a bit and instead introducing an option called something like `AggregateTraces` which would enable the aggregated tracing mode in the datapath and also remove these early traces. This would be a single flag that reduces the amount of notifications from the datapath up to userspace to a maximum of once per packet, but ratelimited to once every N seconds unless there is new information (eg, TCP flags that weren't seen before).

**Comment by tgraf** [comment]
Sounds good

File: test/runtime/monitor.go (hunk 1 above)
Comment (aanm) [reply test/runtime/monitor.go:39]: I got the same feeling  when I read it as a user in `Enable early trace when packets are received from endpoint`

**Review by aanm — CHANGES_REQUESTED** [review]

**Comment by joestringer** [comment]
Closing in favour of #4828.

**Pull request closed without merging**